pkg/operation/botanist/cleanup.go

```diff
@@ -16,209 +16,222 @@ package botanist
+	NamespaceCheckSelector = NamespaceDeleteSelector
+
+	// APIServiceDeleteSelector is the delete selector for APIServices.
+	APIServiceDeleteSelector = &client.ListOptions{
+		LabelSelector: NotSystemComponentSelector.Add(NotKubeAggregatorAutoManaged),
+	}
+	// APIServiceCheckSelector is the check selector for APIServices.
+	APIServiceCheckSelector = APIServiceDeleteSelector
+
+	// CronJobDeleteSelector is the delete selector for CronJobs.
+	CronJobDeleteSelector = &NotSystemComponentListOptions
+	// CronJobCheckSelector is the check selector for CronJobs.
+	CronJobCheckSelector = CronJobDeleteSelector
+
+	// IngressDeleteSelector is the delete selector for Ingresses.
+	IngressDeleteSelector = &NotSystemComponentListOptions
+	// IngressCheckSelector is the check selector for Ingresses.
+	IngressCheckSelector = IngressDeleteSelector
+
+	// JobDeleteSelector is the delete selector for Jobs.
+	JobDeleteSelector = &NotSystemComponentListOptions
+	// JobCheckSelector is the check selector for Jobs.
+	JobCheckSelector = JobDeleteSelector
+
+	// PodDeleteSelector is the delete selector for Pods.
+	PodDeleteSelector = &NotSystemComponentListOptions
+	// PodCheckSelector is the check selector for Pods.
+	PodCheckSelector = PodDeleteSelector
+
+	// ReplicaSetDeleteSelector is the delete selector for ReplicaSets.
+	ReplicaSetDeleteSelector = &NotSystemComponentListOptions
+	// ReplicaSetCheckSelector is the check selector for ReplicaSets.
+	ReplicaSetCheckSelector = ReplicaSetDeleteSelector
+
+	// ReplicationControllerDeleteSelector is the delete selector for ReplicationControllers.
+	ReplicationControllerDeleteSelector = &NotSystemComponentListOptions
+	// ReplicationControllerCheckSelector is the check selector for ReplicationControllers.
+	ReplicationControllerCheckSelector = ReplicationControllerDeleteSelector
+
+	// PersistentVolumeClaimDeleteSelector is the delete selector for PersistentVolumeClaims.
+	PersistentVolumeClaimDeleteSelector = &NotSystemComponentListOptions
+	// PersistentVolumeClaimCheckSelector is the check selector for PersistentVolumeClaims.
+	PersistentVolumeClaimCheckSelector = PersistentVolumeClaimDeleteSelector
+)
 
-		kubernetes.Namespaces: {
-			metav1.NamespacePublic:    true,
-			metav1.NamespaceSystem:    true,
-			metav1.NamespaceDefault:   true,
-			corev1.NamespaceNodeLease: true,
+func cleanResourceFn(c client.Client, deleteSelector, checkSelector *client.ListOptions, list runtime.Object) flow.TaskFn {
+	return flow.SequentialSliced(10*time.Minute,
+		func(ctx context.Context) error {
+			return RetryCleanMatchingUntil(
+				ctx,
+				DefaultInterval,
+				c,
+				deleteSelector,
+				checkSelector,
+				list,
+				false,
+				client.GracePeriodSeconds(60),
+				client.PropagationPolicy(metav1.DeletePropagationForeground),
+			)
 		},
-		kubernetes.Services: {
-			fmt.Sprintf("%s/kubernetes", metav1.NamespaceDefault): true,
+		func(ctx context.Context) error {
```

Comment:
I think we're already doing questionable things by force-deleting APIServices and CRDs - just imagine you have a CRD that hides some very expensive resource below it and we're removing it without finalization to happen. All this is just a trade-off to reduce maintenance effort we have by manually removing such clusters.

However, I agree that this is a behavioral change and I don't want to cause behavioral changes (yet). I've hidden the force deletion behind a flag now so that we still can chose to enable it for other resources in the future (depending on future effort).